documentation/general/dash-sai-pipeline-packet-flow.md

```diff
@@ -0,0 +1,799 @@
+
+At high-level, the metadata bus is a set of fields that being carried all the way through the pipeline along with the packet. It contains:
+
+- The information from the original packet, such as encap information.
+- The information from each matched entry and related data structures, e.g., when a VNET mapping entry is matched, we will publish the information from VNET to the metadata bus. 
+- The routing action and its inline parameters from the matched entry.
+
+Implementation-wise, this is similar to Packet Header Vector or Bus in NPL.
+
+### 5.2. Direction Lookup
+
+In DASH-SAI pipeline, traffic are split into 2 directions: `inbound` and `outbound`. Each direction has its own pipeline (see pipeline overview above). When a new packet arrives, we will assign a direction to the packet, then process the packet in the corresponding pipeline. This ensures us to match the flow and transform the packet in the right way.
+
+### 5.3. Pipeline Lookup
+
+DASH supports multi-tenancy model for traffic handling. A single device can have multiple pipelines, and each pipeline is used to handle traffic for a specific tenant. When a packet arrives, besides direction lookup, we also need pipeline lookup to determine which pipeline to use for processing the packet.
+
+For example, if we like to implement a VM NIC with, then we can model one pipeline as one VM NIC, then use the inner packet MAC to find the pipeline. However, if we would like to implement a load balancer, we can use a DASH-SAI pipeline to represent a load balancer instance and use the Public IP to find the SAI pipeline.
+
+```mermaid
+flowchart TD
+    subgraph ENI
+        EL[ENI Lookup]
+
+        EP0[ENI 0 Pipeline]
+        EP1[ENI 1 Pipeline]
+        EP2[ENI 2 Pipeline]
+    end
+
+    subgraph LoadBalancer
+        VL[Vip Lookup]
+
+        LBP0[LB 0 Pipeline]
+        LBP1[LB 1 Pipeline]
+        LBP2[LB 2 Pipeline]
+    end
+
+    EL --> |ENI 0| EP0
+    EL --> |ENI 1| EP1
+    EL --> |ENI 2| EP2
+
+    VL --> |LB 0| LBP0
+    VL --> |LB 1| LBP1
+    VL --> |LB 2| LBP2
+```
+
+A pipeline can also define its initial matching stages, which will be used for starting processing the packets if no flow is matched.
+
+> NOTE:
+> 
+> DASH-SAI pipeline is a logical concept. It doesn't have to be 1 to 1 mapping to a physical ASIC pipeline. The underlying implementation can be as simple as a metadata field update with a specific value, when the entry getting matched, e.g., `ENI = Inner Source/Destination MAC`.
+
+### 5.4. Packet Decap
+
+If a pipeline is found, before processing the packets, all outer encaps will be decap'ed, and with key information saved in metadata bus, such as encap type, source IP and VNI / GRE Key, exposing the inner most packet going through the pipeline. This simplifies the flow matching logic and also allow us to create the reverse flow properly.
```

Comment:
Can you clarify how to decap all outer encaps? I'm thinking we probably don't want to simply parse headers until there are no more tunnels found. What if the customer traffic is actually a tunnel they created?

Reply:
yea, I was thinking about adding these details but also afraid going too far on this XD. But here is my thoughts:

## Problem of preserving the encaps

Today, the encaps of incoming packet (before pipeline processing) will be used for creating the encaps of the reverse flow, so essentially, we have 2 things to answer:

1. How preserving encaps affects the reverse flow creation?
2. How preserving encaps affects the forwarding side tunnel creation?

With the packet flow doc, now we have a clear definition of the packet structure: overlay (inner most) -> underlay -> tunnel0 -> tunnel1 -> .... However, depends on what user wants, they could do something really interesting, for example, removing or updating the underlay, preserve the tunnel0 and remove tunnel1. So, we are facing all the combinations of CURD of all the encaps including structure change: after removing underlay, should tunnel0 becomes underlay or should we keep it as tunnel0? This affects how the `staticencap` and `tunnel` action works today.

## Solution

In my thought, since these are essentially 2 questions, then we can model it as 2 independent operations. 

### Reverse tunnel creation

For reverse tunnel creation, Since no matter which encap (nvgre or vxlan) we use, they always have a key inside, so we can build the VNI table to specify what we should do. This is already covered with "stateless decap and stateful decap" section. If a VNI is stateless, then we don't create the reverse encap, otherwise we create one.

```json
{
    "DASH_SAI_VNI_TABLE|12345": { "direction": "outbound", "stateless": false },
    "DASH_SAI_VNI_TABLE|12346": { "direction": "outbound", "stateless": true }
}
```

If automatic reverse tunnel creation is not enough, we can also use `reverse_tunnel` to create a dedicated one.

### Forwarding tunnel creation

Although the encap is decaped, but all the encap information is preserved in the metadata bus. Hence, we can easily recreate them in anyway we want by introducing a new action, say: `tunnel_from_encap` (sorry, the name might look silly), and defined as below:

- Parameters from action:
  - Source: "underlay|tunnel0|tunnel1|..."
  - Target: "underlay|tunnel0|tunnel1|..."
- Parameters from metadata:
  - `tunnel_from_encap_(underlay|tunnel0|tunnel1)_sip`: overrides the source ip in the specified target encap
  - `tunnel_from_encap_(underlay|tunnel0|tunnel1)_dip`: overrides the destination ip in the specified target encap
- Action:
  - Enable the target encap by copying the information from the source encap.
  - When overrides exists, uses the overrided one.

If this makes sense, I will add them into the doc.

Reply:
Updated the doc according. Please feel free to take another look, Bud! : D

Reply:
Hi Riff, you wrote "Today, the encaps of incoming packet (before pipeline processing) will be used for creating the encaps of the reverse flow".  I'm not sure if this is true with the dash behavior that's been defined so far. 

In regard to the double decap, I think you are saying run the pipeline twice? After the first tunnel header is decapped, run the pipeline and the LPM/actions etc will direct us to run the pipeline again (and thus trigger another decap)? If I have this wrong please correct me.

I have a higher-level comment sort of related to this and the rest of the doc. I see you are defining the dash pipeline to make it more generic and modular. But this is not the approach that has been taken up until now. Maybe the use cases for DASH have expanded? I'm a little bit concerned about the effort it will take to implement what you're proposing/defining.

Reply:
> Hi Riff, you wrote "Today, the encaps of incoming packet (before pipeline processing) will be used for creating the encaps of the reverse flow". I'm not sure if this is true with the dash behavior that's been defined so far.

Hi Bud, feel free to correct me, if my understanding is not right. 

I believe what you are referring to is this: "For the inbound direction, after LPM/ACL lookup, the pipeline uses the underlay_ip as specified in the ENI table to Vxlan encapsulate the packet.". 

In my understanding, this is right when no flow is created on the inbound side. But if we allow outbound direction creating a flow and let the flow to handle the inbound packet, then this underlay_ip will not be used anymore, since the transformation should be coming from the flow. This maps to how fast path in DPU/ASIC. So, we need to get the IP and stored it into the reverse flow (for inbound side).

The simple idea is to use the underlay_ip in reverse flow. But if source VM using multiple VTEP as ECMP group, it won't be handled well, because it is preferred to use whichever VTEP that sends the traffic to return the packet. Additionally, if the ENI is used for modeling other things which handles multiple sources, such as load balancer or other gateways, the source IP will not be a single IP or can be configured on the ENI. So it is better for us to learn it from the original packet.

> In regard to the double decap, I think you are saying run the pipeline twice? After the first tunnel header is decapped, run the pipeline and the LPM/actions etc will direct us to run the pipeline again (and thus trigger another decap)? If I have this wrong please correct me.

Not really. We don't recirculate the packet, but let the parser to handle certain level of the encaps. What you are saying is actually a very good one. I believe the confusion is caused by we didn't have the maximum number of encaps defined. If this solves your concern, I will update the doc.

> I have a higher-level comment sort of related to this and the rest of the doc. I see you are defining the dash pipeline to make it more generic and modular. But this is not the approach that has been taken up until now. Maybe the use cases for DASH have expanded? I'm a little bit concerned about the effort it will take to implement what you're proposing/defining.

What you are feeling is right. I think DASH is doing good job modeling certain VNET scenarios. And it has all the corner stones to be flexible and modular, but it is not using them at all. For example, routing action behavior changes based on routing types. Routing types is used as the real minimum building block, making routing action almost useless. These things limit the usage of DASH a lot.

If you look at the proposal and ignore the inbound pipeline and multi-stage update, you will see the object model is not changed much at all (VNI table is introduced for handling multi-layer encap and multi-VNI cases, so it doesn't really count). But you might already feel the pipeline being much more flexible and can do a lot of more things.

For your concern, yes, I share the same concern as well. There are certain fundamental things being changed in this proposal, such as moving from programming routing type to programming routing action. It could increase the underlying implementation complexity. But on the other hand, with a simple mind model, the complexity could be highly reduced as well. No more corner cases, no more special handling, everything just moves as you see it.

So, although I have the same concern, I still would like to get this doc and proposal out, discussed and gather the feedbacks. And after this, we will be able to have a clear definition on how the pipeline works instead of having many actions being vague and do whichever they like under different conditions.

Hope this solves your question!

Reply:
the maximum number of decap is added into the doc.

Reply:
Hi Riff,
Just writing down my comment from the last meeting regarding handling packets that arrive with two tunnel encaps. IMO always decapping the two tunnel headers is not a good solution. If the outer tunnel header is for DASH and the inner tunnel is a customer tunnel, then we would probably drop the packet. (The customer tunnel wouldn't match any valid DASH tunnel.) This basically means that DASH doesn't let customers use tunnels, which is kind of a severe limitation.

Not sure what the solution is. I think ideally you'd like to check the outer tunnel to know if a second tunnel is expected. For instance, if a mux can put a second encap on on a packet, then maybe the VNI/tunnel_id the mux applies is known to DASH orchestration. And the DASH direction_lookup table (which has the tunnel_id for a key) includes a double_encap flag. This would be set to true for any tunnel from a mux. At packet forwarding time, we do the direction_table lookup. If this flag is not set in the lookup result then we decap only the outer tunnel. If the flag is set, then not only do we decap a second tunnel, but we drop the packet if a second tunnel is not present. I know there are issues with implementing this with the P4 parser -- I'm only suggesting a logical solution.

Reply:
Thanks a lot for capturing this Bud! I have updated the doc with multi-layer encap handling for this problem, which should be doable with P4 with `value_set` feature in parser. Do you mind to help take another look and see how it looks? : D

On high level, it works with 2 things:

1. use `final_encap` flag in VNI, so we can tell the parser that there is no more encap after this.
2. use some code to fix or extract the encap info in metadata after parsing, similar to what DASH is doing today.